### PR TITLE
Feature/25 add skip to location

### PR DESCRIPTION
**Pull request by codiebeulaine**
@hodgestar ready for review

### CODE REVIEW

File: lib/location.js [hunk 1]
```diff
@@ -261,10 +270,19 @@ var LocationState = State.extend(function (self, name, opts) {
...
+        };
+    }
+    else
+    {
+        self.generate_footer = function() {
         return '\n' + 'n. '+self.next_text + '\n' + 'p. '+self.previous_text;
-
-    };
+        };
+    }
```
Comment (hodgestar) [inline lib/location.js]: I would avoid have separate definitions of `generate_footer` and just do something like:

File: test/test_location.js [hunk 2]
```diff
@@ -537,4 +537,5 @@ describe('states.location', function() {
         });
 
     });
+
```
Comment (hodgestar) [inline test/test_location.js]: I would revert this change if there is no need to touch this file.

File: test/test_skip.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+            return tester
+                .start()
+                .check.interaction({
+                    state:'states:test',
+                    reply:'Custom location question',
+                    })
+                .run();
+        });
+
+        it('should give a list of locations when the user enters an address',
```
Comment (hodgestar) [inline test/test_skip.js]: I think you only need this test (the others just test things that are already tested elsewhere) and it should live in the same file as the other location tests. You can activate skipping by doing:
```
tester.data.opts.skip_text = 'Skip';
```
at the start of the test.

**Comment by hodgestar** [comment]
Couple of minor comments. Otherwise looks good.

**Comment by codiebeulaine** [comment]
@hodgestar made those changes, I used the skip test I added, is this okay? They all live in the same file now

**Comment by codiebeulaine** [comment]
@hodgestar also, added the retry option. Having trouble reading the metadata though

File: lib/location.js [hunk 4]
```diff
@@ -261,9 +275,16 @@ var LocationState = State.extend(function (self, name, opts) {
         return '\n' + number + '. ' + address.label;
     };
 
-    self.generate_footer = function() {
-        return '\n' + 'n. '+self.next_text + '\n' + 'p. '+self.previous_text;
-
+    self.generate_footer = function(content) {
+        var options = [];
+        if (self.metadata.current_page >= self.metadata.pages.length && self.retry_text){
```
Comment (hodgestar) [inline lib/location.js]: What is the `current_page > pages.length` check for?

Comment (hodgestar) [inline lib/location.js]: Should `r. Retry` be after the next and previous options?

File: test/test_location.js [hunk 5]
```diff
@@ -189,6 +200,20 @@ describe('states.location', function() {
                 .run();
         });
 
+        it('should go back to initial page if retry is selected',
+        function() {
+            tester.data.opts.retry_text = 'Retry';
+            return tester
+                .inputs("Friend Street", 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n',
+                    'n', 'n', 'n', 'n', 'n', 'n', 'n', 'r')
```
Comment (hodgestar) [inline test/test_location.js]: We can maybe skip some / all of the `'n'`s after my comment about generate footer has been addressed.

**Comment by hodgestar** [comment]
Tests look good now! Thank you. :)

**Comment by hodgestar** [comment]
Currently I think if one presses `s` or `r` they will work even if the relevant options aren't displayed in the footer? Perhaps we should only add the corresponding handlers if `skip_text` and `retry_text` are set?

File: lib/location.js (hunk 4 above)
Comment (codiebeulaine) [inline lib/location.js]: r. Retry should replace the n. Next on the last page only. The current_page > pages.length check is to see whether the last page is being displayed so that the retry option can replace the next option

Comment (hodgestar) [inline lib/location.js]: What you describe is sensible, but that's not what the code currently does. 

I would vote for the `r. Retry` to be on every page if it's active (otherwise one can be stuck on page 3/5 wanting to retry?). Thoughts?

File: test/test_location.js (hunk 5 above)
Comment (codiebeulaine) [inline test/test_location.js]: sounds good, will do that :)

File: lib/location.js (hunk 4 above)
Comment (codiebeulaine) [inline lib/location.js]: This was my initial thought as well, because otherwise your only way out is the skip option or pressing next a trillion times, but I didn't want to question the ticket

Comment (hodgestar) [inline lib/location.js]: Always question the ticket! :)

Comment (codiebeulaine) [inline lib/location.js]: haha, cool. Will make these changes as well :)

Comment (codiebeulaine) [inline lib/location.js]: @hodgestar still getting the same undefined error. Tried researching it but not quite sure what I'm doing wrong

**Comment by codiebeulaine** [comment]
@hodgestar ready for review :) again haha

**Comment by hodgestar** [comment]
Looks good!

We need two extra tests that test that `Skip` and `Retry` are displayed in the menus (currently we test that pressing `s` and `r` does the right thing).

File: lib/location.js [hunk 6]
```diff
@@ -166,13 +170,22 @@ var LocationState = State.extend(function (self, name, opts) {
...
+        var choices = {
             n: self.handlers.refinequestion.next,
             p: self.handlers.refinequestion.prev
-        }[content];
+        };
+
+        if (self.skip_text){
+            choices.s = self.handlers.skip;
+        }
+        if (self.retry_text){
```
Comment (hodgestar) [inline lib/location.js]: Very nitpicky thing -- we don't have a space between `){` in these two, but we do elsewhere.

File: test/test_location.js [hunk 7]
```diff
@@ -369,6 +392,44 @@ describe('states.location', function() {
             .run();
         });
 
+        it('should display skip option when skip text not null',
+        function() {
+            tester.data.opts.options_per_page = 1;
+            tester.data.opts.skip_text = 'Skip';
+            return tester
+            .inputs("Friend Street")
```
Comment (hodgestar) [inline test/test_location.js]: Other nitpick -- `.inputs` isn't indented like it is in other tests.

Comment (codiebeulaine) [inline test/test_location.js]: @hodgestar not quite sure what you mean? Do you mean it is not necessary?

Comment (hodgestar) [inline test/test_location.js]: I mean:
```
return tester
    .inputs("Friend Street")
    .check.interaction({
        ...
     })
     .run();
```
would be more consistent.

Comment (codiebeulaine) [inline test/test_location.js]: I read that wrong, I read intended instead on indented. Thanks :) ready for review once again :)

**Comment by hodgestar** [comment]
:+1:
